### DISCUSSION TITLE
Integration in System.Data.SQLite (.NET)

### DISCUSSION THREAD

**Discussion by MarcelVersteeg** [utelle/SQLite3MultipleCiphers#45] [General]
Can you give a hint on how to use this project in the System.Data.SQLite assemblies/distribution so the ciphers can be used in .NET development?

**Comment by utelle** [comment]
I suppose you refer to [System.Data.SQLite](https://system.data.sqlite.org). As far as I know its implementation is based on the SQLite amalgamation source code. So the best approach will certainly be to replace the SQLite amalgamation by the _SQLite3 Multiple Ciphers_ amalgamation (which is made available with each release or which can be created with a script from the original sources). However, additionally it will be most likely necessary to adjust the code in `SQLite.Interop/src/generic/interop.c` of the _System.Data.SQLite_ distribution. And it will be necessary to adjust the list of compile time symbols. Maybe the postings in [wxSQLite3 issue #24](https://github.com/utelle/wxsqlite3/issues/24) are also useful.

To enable the methods/functions to set the encryption key it will be necessary to define the symbol `INTEROP_CODEC`. However, this has an unwanted side effect: the code in `SQLite.Interop/src/win/crypt.c` is included in `interop.c` - this must be avoided!

Additionally, it might be useful to make the additional API functions fo _SQLite3 Multiple Ciphers_ available (in the same way as the functions `sqlite3_key` and `sqlite3_rekey`), but it is not absolutely necessary, because the cipher scheme can be selected and configured via PRAGMA commands or URI parameters.

An alternative might be _Microsoft.Data.SQLite_. There is for example a blog post by _Brice Lambson_, [Encryption in Microsoft.Data.Sqlite](https://www.bricelam.net/2016/06/13/sqlite-encryption.html). However, I have no idea whether integrating _SQLite3 Multiple Ciphers_ into _Microsoft.Data.SQLite_ is easier to accomplish.

**Reply by shamork** [reply]
Integrating SQLite3 Multiple Ciphers into Microsoft.Data.SQLite IS easier to accomplish.
I've make it working, steps in https://github.com/shamork/SQLitePCLRaw.multipleciphers

**Reply by utelle** [reply]
> Integrating SQLite3 Multiple Ciphers into Microsoft.Data.SQLite IS easier to accomplish. I've make it working, steps in https://github.com/shamork/SQLitePCLRaw.multipleciphers

Thanks for providing a description of how to use _SQLite Multiple Ciphers_ with _Microsoft.Data.SQLite_. For sure, this will be very helpful for developers seeking support for _SQLite3 Multiple Ciphers_ in the .Net world.

**Comment by utelle** [comment]
Based on the latest officially released _System.Data.SQLite_ code (namely version 1.0.113.0) I checked which modifications might be required to integrate _SQLite3 Multiple Ciphers_ in _System.Data.SQLite_.

A few minor adjustments to the code of _SQLite3 Multiple Ciphers_ are required to make it compile with the solution files coming with _System.Data.SQLite_. These adjustments were applied in commit 5013599d6a598e54f36ae15e8b9d5e453bfe3965.

Additionally, the property files as well as the project files of _System.Data.SQLite_ need minor adjustments - adding compile time symbols in the property files, removing several source files which cause conflicts from the project files.

As already mentioned mainly the source file `SQLite.Interop/src/generic/interop.c` has to be adjusted (removing `#include` statements for SQLite extensions which are already included in the _SQLite3 Multiple Ciphers_ amalgamation in the first place).

**Comment by MarcelVersteeg** [comment]
Thank you for the information (and the changes to your code).

I will look into it and see if I can get it to compile. I might consider creating a project out of it on GitHub, as others might also be interested in it.

**Comment by utelle** [comment]
Looking at the [System.Data.SQLite timeline](http://system.data.sqlite.org/index.html/timeline?n=20&y=ci) I get the impression that the release of _System.Data.SQLite_ version **1.0.114.0** will happen in the near future. Most likely integration of _SQLite3 Multiple Ciphers_ into _System.Data.SQLite_ will be easier to accomplish with that new version, because it will be already adapted to the latest SQLite version. So I would recommend to wait a few days (or to use a recent development snapshot).

**Comment by kroimon** [comment]
I would be very interested in any solution allowing me to use _SQLite3 Multiple Ciphers_ in .NET as I need to implement support for reading databases encrypted by wxsqlite3 (using AES128 CBC) in a .NET application.

I had a quick look around and I feel like this might be easier to do with _Microsoft.Data.Sqlite_ as it is using [SQLitePCLRaw](https://github.com/ericsink/SQLitePCL.raw) which is specifically built to use a variety of sqlite libraries.
@utelle Do you think there are any reasons this might not work with just using the _SQLite3 Multiple Ciphers_ release binaries directly with _SQLitePCLRaw_?

**Comment by utelle** [comment]
@kroimon 
>Do you think there are any reasons this might not work with just using the SQLite3 Multiple Ciphers release binaries directly with SQLitePCLRaw?

Since SQLitePCLRaw seems to use SQLite DLLs, I see no reason why it should not work. The _SQLite3 Multiple Ciphers_ DLLs provided with the releases support the same API as the original SQLite DLL. 

There are a few additional functions in the _SQLite3 Multiple Ciphers_ interface, but they are not needed to make use of the _SQLite3 Multiple Ciphers_ DLL, because the encryption extension can be configured via SQL pragma statements or via URI parameters (on opening a SQLite database).

That is, renaming the DLL coming with _SQLite3 Multiple Ciphers_ releases to **sqlite3.dll** should allow you to simply replace the DLL used by _SQLitePCLRaw_.

However, in respect to security using SQLite DLLs is not ideal, because an attacker could easily intercept calls into the DLL.

**Comment by haiduong87** [comment]
Hi @MarcelVersteeg,

Did you success making your own System.Data.SQlite?
I'm in this trouble right now. 
I try to follow the conversation, but the source code of both projects have some recent updates...

@utelle ,
Can you help me here (I can't find any "amalgamation" file in this source code, https://system.data.sqlite.org/index.html/doc/trunk/www/downloads.wiki - sqlite-netFx-source-1.0.114.0.zip (7.67 MiB))

**Comment by utelle** [comment]
> Can you help me here (I can't find any "amalgamation" file in this source code, https://system.data.sqlite.org/index.html/doc/trunk/www/downloads.wiki - sqlite-netFx-source-1.0.114.0.zip (7.67 MiB))

If you unzip the archive `sqlite-netFx-source-1.0.114.0.zip` you will find therein subdirectory `SQLite.Interop/src/core`. This subdirectory contains the files `sqlite3.c` and `sqlite3.h`, which constitute the _SQLite amalgamation_. The first step would be to replace those 2 files by the _SQLite3 Multiple Ciphers_ amalgamation (`sqlite3mc_amalgamation.c` and `sqlite3mc_amalgamation.h`). However, this is not enough. In a second step you need to adjust the file `SQLite.Interop/src/generic/interop.c` to resolve compilation conflicts. Additionally, it is most likely necessary to adjust the solution/project files by adding or changing some preprocessor symbols.

**Comment by utelle** [comment]
I will convert this issue into a discussion, because the **SQLite3 Multiple Ciphers** amalgamation source code can be used as a drop-in replacement for the **SQLite3** amalgamation source code used in _System.Data.SQLite_. As described in my previous post full integration in _System.Data.SQLite_ is then only a matter of adjusting the file `SQLite.Interop/src/generic/interop.c` and the set of used preprocessor symbol definitions. However, this is not a task of **SQLite3 Multiple Ciphers**.
